src/store/mqtt/actions.ts

```diff
@@ -39,81 +51,113 @@ export function connect() {
+    client.on('close', () => dispatch({ type: constants.CLOSE }));
+    client.on('offline', () => dispatch({ type: constants.OFFLINE }));
+    client.on('reconnect', () => dispatch(reconnect()));
+    client.on('message', (topic: string, message: MqttMessage) =>
+      dispatch(handleMessage(topic, message)));
+
+    dispatch({ type: constants.CONNECT_END });
+  };
+}
+
+
+// This is one of the most complex things in the code,
+// the main purpose of this logic is to try to auto-reconnect after disconnecting.
+// BUT, we cannot let mqtt.js handle this automatically for us, because our password
+// (jwt) changes every few minutes (because that's why the session get's renewed and extended)
+// so need to hijack the reconnection mechanism provided by mqtt.js, abort it (disconnect) and
+// re initiate the connection ourselves, by dispatching(connect) recursively.
+export function reconnect() {
+  return async function(dispatch: Dispatch, getState: GetState) {
+    const state = getState().mqtt;
+    if (state.disconnecting) {
+      return;
+    }
+
+    dispatch({ type: constants.RECONNECT_START });
+    // save state.mqtt.data
+    const subscriptions = cloneDeep(state.data);
+    client.end(true);
+    client = null;
+    dispatch({ type: constants.NOT_CONNECTED });
+
+    // Exponential backoff
+    let delay = 500 * Math.pow(2, state.reconnectCount);
+    if (delay > MAX_DELAY) {
+      delay = MAX_DELAY;
+    }
+    setTimeout(
+      async () => {
+        try {
+          await dispatch(connect());
+        } catch (err) {
+          dispatch(restoreSubscriptionState(subscriptions));
+          dispatch({ type: constants.RECONNECT_END });
+          // This recursive call only happens when a long disconnection happens
+          dispatch(reconnect());
           return;
         }
-      });
-      // restore state.mqtt.data
-      dispatch(restoreSubscriptionState(subscriptions));
-    };
-
-    Object.keys(handlers)
-      .map((eventName) => [eventName, handlers[eventName]])
-      .forEach(([eventName, handler]) => {
-        client.on(eventName, handler);
-      });
+        // re subscribe mqtt
+        client.subscribe(Object.keys(subscriptions), (error, granted) => {
+          if (error) {
+            dispatch(setError(error.message));
+            console.log(`MQTT: something went wrong while subscribing`, error);
+            return;
+          }
+        });
+        // restore state.mqtt.data
+        dispatch(restoreSubscriptionState(subscriptions));
+        dispatch({ type: constants.RECONNECT_END });
+      },
+      delay,
+    );
   };
 }
 
 export function disconnect() {
   return async function(dispatch: Dispatch, getState: GetState) {
+    dispatch({ type: constants.DISCONNECT_START });
     const state = getState();
-    if (!state.mqtt.connected) {
-      console.log(`MQTT ERROR: attempting to disconnect while already disconnected`);
+    if (!client || !state.mqtt.connected) {
```

Comment:
would love to switch to the actual mqtt client's understanding of whether or not we are connected, but I put that in an issue on the original repo: https://github.com/ionia-corporation/xively-react-boilerplate/issues